pytket/extensions/cutensornet/tnstate/ttn.py

```diff
@@ -0,0 +1,854 @@
+
+        Notes:
+            After applying this function, ``self`` will contain the projected
+            state over the non-postselected qubits.
+
+            The resulting state has been normalised.
+
+        Args:
+            qubit_outcomes: A dictionary mapping a subset of qubits to their
+                desired outcome value (either ``0`` or ``1``).
+
+        Returns:
+            The probability of this postselection to occur in a measurement.
+
+        Raises:
+            ValueError: If a key in ``qubit_outcomes`` is not a qubit in the state.
+            ValueError: If a value in ``qubit_outcomes`` is other than ``0`` or ``1``.
+            ValueError: If all of the qubits in the state are being postselected.
+                Instead, you may wish to use ``get_amplitude()``.
+        """
+        raise NotImplementedError(f"Method not implemented in {type(self).__name__}.")
+
+    def expectation_value(self, pauli_string: QubitPauliString) -> float:
+        """Obtains the expectation value of the Pauli string observable.
+
+        Args:
+            pauli_string: A pytket object representing a tensor product of Paulis.
+
+        Returns:
+            The expectation value.
+
+        Raises:
+            ValueError: If a key in ``pauli_string`` is not a qubit in the state.
+        """
+        raise NotImplementedError(f"Method not implemented in {type(self).__name__}.")
+
+    def get_fidelity(self) -> float:
+        """Returns the current fidelity of the state."""
+        return self.fidelity
+
+    def get_statevector(self) -> np.ndarray:
+        """Returns the statevector represented by the TTN, with qubits ordered
+        in Increasing Lexicographic Order (ILO).
+        Raises:
+            ValueError: If there are no qubits left in the TTN.
+        """
+        if len(self.get_qubits()) == 0:
+            raise ValueError("There are no qubits left in this TTN.")
+
+        # Create the interleaved representation with all tensors
+        interleaved_rep = self.get_interleaved_representation()
+
+        # Specify the output bond IDs in ILO order
+        output_bonds = []
+        for q in sorted(self.get_qubits()):
+            output_bonds.append(str(q))
+        interleaved_rep.append(output_bonds)
+
+        # Contract
+        result_tensor = cq.contract(
+            *interleaved_rep,
+            options={"handle": self._lib.handle, "device_id": self._lib.device_id},
+            optimize={"samples": 1},  # There is little to no optimisation to be done
+        )
+
+        # Convert to numpy vector and flatten
+        statevector: np.ndarray = cp.asnumpy(result_tensor).flatten()
+        return statevector
+
+    def get_amplitude(self, state: int) -> complex:
+        """Returns the amplitude of the chosen computational state.
+
+        Notes:
+            The result is equivalent to ``self.get_statevector[b]``, but this method
+            is faster when querying a single amplitude.

```

Comment:
I wonder if it IS faster though... all the steps seem either the same or very comparable in cost. Not that I am against of this method, ofc.

Reply:
Oh, but here you contract to a scalar, while in `get_statevector` - to a vector - I guess it is cheaper at least memory-wise then.

Reply:
Indeed, and not only memory wise, this allows for a contraction path that has much smaller intermediate tensors, so it should actually be _much_ faster. In asymptotic terms, `get_amplitude` is `poly(chi)` whereas `get_statevector` is `O(2^n)` where `n` is the number of qubits.